### ISSUE TITLE
Ormolu completely deletes misplaced Haddock comments

### ISSUE THREAD

**Issue by lrworth** [tweag/ormolu#474]
Formatting:
```
test = undefined
  where
    -- | Comment
    a = undefined
```
results in:
```
test = undefined
  where
    a = undefined
```

**Comment by mrkkrp** [comment]
The `-- |` style is for Haddocks, and they cannot be attached to non-top-level functions. So what happens here is that GHC parser that we use discards the Haddock completely because it's not a normal `--` comment and is not a valid Haddock either.

If it is a bug, it is a bug in GHC parser, but I'm not sure what it should be doing instead in this case. In short, it is better not to use Haddock syntax such as `-- |` where it is not supposed to be used.

Note that normal comment gets preserved:

```haskell
test = undefined
  where
    -- Comment
    a = undefined
```

**Comment by neongreen** [comment]
I would say it's perfectly sensible to use Haddock syntax.

`-- |` does seem a bit dubious, but `-- ^` for function arguments is both reasonable and accepted by GHC. For instance, Ormolu formats this just fine:

```haskell
a = undefined
  where
    foo :: Int -- ^ a
        -> Bool -- ^ b
    foo = undefined
```

I filed an issue: https://gitlab.haskell.org/ghc/ghc/issues/17538. I think it should be classified as `bug-upstream`.

**Comment by lrworth** [comment]
It was alarming to find things like that after running ormolu over our codebase for the first time. If there is a syntax problem it should be reported instead of silently deleted.

**Comment by srid** [comment]
Here's another case:

```haskell
instance Show Foo where
  -- | This will be removed by Ormolu, unless we delete the "|" 
  show = undefined
```

**Comment by mrkkrp** [comment]
Happens with `-- *` lines too, see #524.

**Comment by amesgen** [comment]
I think that we will be able to fix this when we upgrade to ghc-lib-parser-9.2 🎉 . The dumped AST of

```haskell
test = undefined
  where
    -- | misplaced
    a = undefined
```

in GHC 9.2-alpha2 contains "misplaced":

<details>
<summary>Click to expand, and search for "misplaced"</summary>

```
(L
 { <input>:1:1 }
 (HsModule
  (EpAnn
   (Anchor
    { <input>:1:1 }
    (UnchangedAnchor))
   (AnnsModule
    []
    (AnnList
     (Nothing)
     (Nothing)
     (Nothing)
     []
     []))
   (EpaCommentsBalanced
    []
    [(L
      (Anchor
       { <input>:5:1 }
       (UnchangedAnchor))
      (EpaComment
       (EpaEofComment)
       { <input>:5:1 }))]))
  (VirtualBraces
   (1))
  (Nothing)
  (Nothing)
  []
  [(L
    (SrcSpanAnn (EpAnn
                 (Anchor
                  { <input>:(1,1)-(4,17) }
                  (UnchangedAnchor))
                 (AnnListItem
                  [])
                 (EpaComments
                  [])) { <input>:(1,1)-(4,17) })
    (ValD
     (NoExtField)
     (FunBind
      (NoExtField)
      (L
       (SrcSpanAnn (EpAnnNotUsed) { <input>:1:1-4 })
       (Unqual
        {OccName: test}))
      (MG
       (NoExtField)
       (L
        (SrcSpanAnn (EpAnnNotUsed) { <input>:(1,1)-(4,17) })
        [(L
          (SrcSpanAnn (EpAnnNotUsed) { <input>:(1,1)-(4,17) })
          (Match
           (EpAnn
            (Anchor
             { <input>:(1,1)-(4,17) }
             (UnchangedAnchor))
            []
            (EpaComments
             []))
           (FunRhs
            (L
             (SrcSpanAnn (EpAnnNotUsed) { <input>:1:1-4 })
             (Unqual
              {OccName: test}))
            (Prefix)
            (NoSrcStrict))
           []
           (GRHSs
            (NoExtField)
            [(L
              { <input>:(1,6)-(4,17) }
              (GRHS
               (EpAnn
                (Anchor
                 { <input>:(1,6)-(4,17) }
                 (UnchangedAnchor))
                (GrhsAnn
                 (Nothing)
                 (AddEpAnn AnnEqual (EpaSpan { <input>:1:6 })))
                (EpaComments
                 [(L
                   (Anchor
                    { <input>:3:5-18 }
                    (UnchangedAnchor))
                   (EpaComment
                    (EpaDocCommentNext
                     " misplaced")
                    { <input>:2:3-7 }))]))
               []
               (L
                (SrcSpanAnn (EpAnnNotUsed) { <input>:1:8-16 })
                (HsVar
                 (NoExtField)
                 (L
                  (SrcSpanAnn (EpAnnNotUsed) { <input>:1:8-16 })
                  (Unqual
                   {OccName: undefined}))))))]
            (HsValBinds
             (EpAnn
              (Anchor
               { <input>:(2,3)-(4,17) }
               (UnchangedAnchor))
              (AnnList
               (Just
                (Anchor
                 { <input>:4:5-17 }
                 (UnchangedAnchor)))
               (Nothing)
               (Nothing)
               [(AddEpAnn AnnWhere (EpaSpan { <input>:2:3-7 }))]
               [])
              (EpaComments
               []))
             (ValBinds
              (NoAnnSortKey)
              {Bag(LocatedA (HsBind GhcPs)):
               [(L
                 (SrcSpanAnn (EpAnn
                              (Anchor
                               { <input>:4:5-17 }
                               (UnchangedAnchor))
                              (AnnListItem
                               [])
                              (EpaComments
                               [])) { <input>:4:5-17 })
                 (FunBind
                  (NoExtField)
                  (L
                   (SrcSpanAnn (EpAnnNotUsed) { <input>:4:5 })
                   (Unqual
                    {OccName: a}))
                  (MG
                   (NoExtField)
                   (L
                    (SrcSpanAnn (EpAnnNotUsed) { <input>:4:5-17 })
                    [(L
                      (SrcSpanAnn (EpAnnNotUsed) { <input>:4:5-17 })
                      (Match
                       (EpAnn
                        (Anchor
                         { <input>:4:5-17 }
                         (UnchangedAnchor))
                        []
                        (EpaComments
                         []))
                       (FunRhs
                        (L
                         (SrcSpanAnn (EpAnnNotUsed) { <input>:4:5 })
                         (Unqual
                          {OccName: a}))
                        (Prefix)
                        (NoSrcStrict))
                       []
                       (GRHSs
                        (NoExtField)
                        [(L
                          { <input>:4:7-17 }
                          (GRHS
                           (EpAnn
                            (Anchor
                             { <input>:4:7-17 }
                             (UnchangedAnchor))
                            (GrhsAnn
                             (Nothing)
                             (AddEpAnn AnnEqual (EpaSpan { <input>:4:7 })))
                            (EpaComments
                             []))
                           []
                           (L
                            (SrcSpanAnn (EpAnnNotUsed) { <input>:4:9-17 })
                            (HsVar
                             (NoExtField)
                             (L
                              (SrcSpanAnn (EpAnnNotUsed) { <input>:4:9-17 })
                              (Unqual
                               {OccName: undefined}))))))]
                        (EmptyLocalBinds
                         (NoExtField)))))])
                   (FromSource))
                  []))]}
              [])))))])
       (FromSource))
      [])))]
  (Nothing)
  (Nothing)))
```

</details>

**Comment by brandonchinn178** [comment]
FYI when this is fixed, you might want to add a regression test, something like: https://github.com/fourmolu/fourmolu/pull/106/commits/91470f23ac3b1ad49087ee172ff10225d492d485

I'm pretty sure this broke between `ghc-lib-parser-8.10` and `ghc-lib-parser-9.0.1`

**Comment by amesgen** [comment]
> I'm pretty sure this broke between `ghc-lib-parser-8.10` and `ghc-lib-parser-9.0.1`

This issue is from December 2019, so it was there even with `ghc-lib-parser-8.10`:

```
 $ for v in 0.1.4.1 0.2.0.0; do wget -O ormolu-$v https://github.com/tweag/ormolu/releases/download/$v/ormolu-Linux; chmod +x ormolu-$v; done
 $ cat test.hs
test = undefined
  where
    -- | Comment
    a = undefined
$ diff <(./ormolu-0.1.4.1 test.hs) <(./ormolu-0.2.0.0 test.hs)
```

As I wrote in my [comment above](https://github.com/tweag/ormolu/issues/474#issuecomment-889182337), I hope that `ghc-lib-parser-9.2` will enable us to fix this.

**Comment by brandonchinn178** [comment]
can you try with this example?
```hs
test = undefined
  where
    a ::
      -- | Comment
      Int
    a = undefined
```
This keeps the comment with
```
fourmolu 0.3.0.0 UNKNOWN UNKNOWN
using ghc-lib-parser 8.10.5.20210606
```
and breaks when switching to `ghc-lib-parser 9.0.1`. I'm not sure if there was a separate change that got fourmolu to work on `ghc-lib-parser 8.10`.

**Comment by amesgen** [comment]
Indeed:

```
 $ diff <(./ormolu-0.1.4.1 test.hs) <(./ormolu-0.2.0.0 test.hs)
4d3
<       -- | Comment
```

So Ormolu 0.1.4.1 (pre `ghc-lib-parser-9.0`)  also did handle this case correctly. This seem to be a regression in `ghc-lib-parser-9.0`.

**Comment by amesgen** [comment]
But this should also be fixed with `ghc-lib-parser-9.2`: The dumped AST of

```haskell
test = undefined
  where
    a ::
      -- | misplaced
      Int
    a = undefined
```

in GHC 9.2-alpha2 contains "misplaced":

<details>
<summary>Click to expand, and search for "misplaced"</summary>

```
(L
 { <input>:1:1 }
 (HsModule
  (EpAnn
   (Anchor
    { <input>:1:1 }
    (UnchangedAnchor))
   (AnnsModule
    []
    (AnnList
     (Nothing)
     (Nothing)
     (Nothing)
     []
     []))
   (EpaCommentsBalanced
    []
    [(L
      (Anchor
       { <input>:7:1 }
       (UnchangedAnchor))
      (EpaComment
       (EpaEofComment)
       { <input>:7:1 }))]))
  (VirtualBraces
   (1))
  (Nothing)
  (Nothing)
  []
  [(L
    (SrcSpanAnn (EpAnn
                 (Anchor
                  { <input>:(1,1)-(6,17) }
                  (UnchangedAnchor))
                 (AnnListItem
                  [])
                 (EpaComments
                  [])) { <input>:(1,1)-(6,17) })
    (ValD
     (NoExtField)
     (FunBind
      (NoExtField)
      (L
       (SrcSpanAnn (EpAnnNotUsed) { <input>:1:1-4 })
       (Unqual
        {OccName: test}))
      (MG
       (NoExtField)
       (L
        (SrcSpanAnn (EpAnnNotUsed) { <input>:(1,1)-(6,17) })
        [(L
          (SrcSpanAnn (EpAnnNotUsed) { <input>:(1,1)-(6,17) })
          (Match
           (EpAnn
            (Anchor
             { <input>:(1,1)-(6,17) }
             (UnchangedAnchor))
            []
            (EpaComments
             []))
           (FunRhs
            (L
             (SrcSpanAnn (EpAnnNotUsed) { <input>:1:1-4 })
             (Unqual
              {OccName: test}))
            (Prefix)
            (NoSrcStrict))
           []
           (GRHSs
            (NoExtField)
            [(L
              { <input>:(1,6)-(6,17) }
              (GRHS
               (EpAnn
                (Anchor
                 { <input>:(1,6)-(6,17) }
                 (UnchangedAnchor))
                (GrhsAnn
                 (Nothing)
                 (AddEpAnn AnnEqual (EpaSpan { <input>:1:6 })))
                (EpaComments
                 []))
               []
               (L
                (SrcSpanAnn (EpAnnNotUsed) { <input>:1:8-16 })
                (HsVar
                 (NoExtField)
                 (L
                  (SrcSpanAnn (EpAnnNotUsed) { <input>:1:8-16 })
                  (Unqual
                   {OccName: undefined}))))))]
            (HsValBinds
             (EpAnn
              (Anchor
               { <input>:(2,3)-(6,17) }
               (UnchangedAnchor))
              (AnnList
               (Just
                (Anchor
                 { <input>:(3,5)-(6,17) }
                 (UnchangedAnchor)))
               (Nothing)
               (Nothing)
               [(AddEpAnn AnnWhere (EpaSpan { <input>:2:3-7 }))]
               [])
              (EpaComments
               []))
             (ValBinds
              (NoAnnSortKey)
              {Bag(LocatedA (HsBind GhcPs)):
               [(L
                 (SrcSpanAnn (EpAnn
                              (Anchor
                               { <input>:6:5-17 }
                               (UnchangedAnchor))
                              (AnnListItem
                               [])
                              (EpaComments
                               [])) { <input>:6:5-17 })
                 (FunBind
                  (NoExtField)
                  (L
                   (SrcSpanAnn (EpAnnNotUsed) { <input>:6:5 })
                   (Unqual
                    {OccName: a}))
                  (MG
                   (NoExtField)
                   (L
                    (SrcSpanAnn (EpAnnNotUsed) { <input>:6:5-17 })
                    [(L
                      (SrcSpanAnn (EpAnnNotUsed) { <input>:6:5-17 })
                      (Match
                       (EpAnn
                        (Anchor
                         { <input>:6:5-17 }
                         (UnchangedAnchor))
                        []
                        (EpaComments
                         []))
                       (FunRhs
                        (L
                         (SrcSpanAnn (EpAnnNotUsed) { <input>:6:5 })
                         (Unqual
                          {OccName: a}))
                        (Prefix)
                        (NoSrcStrict))
                       []
                       (GRHSs
                        (NoExtField)
                        [(L
                          { <input>:6:7-17 }
                          (GRHS
                           (EpAnn
                            (Anchor
                             { <input>:6:7-17 }
                             (UnchangedAnchor))
                            (GrhsAnn
                             (Nothing)
                             (AddEpAnn AnnEqual (EpaSpan { <input>:6:7 })))
                            (EpaComments
                             []))
                           []
                           (L
                            (SrcSpanAnn (EpAnnNotUsed) { <input>:6:9-17 })
                            (HsVar
                             (NoExtField)
                             (L
                              (SrcSpanAnn (EpAnnNotUsed) { <input>:6:9-17 })
                              (Unqual
                               {OccName: undefined}))))))]
                        (EmptyLocalBinds
                         (NoExtField)))))])
                   (FromSource))
                  []))]}
              [(L
                (SrcSpanAnn (EpAnnNotUsed) { <input>:(3,5)-(5,9) })
                (TypeSig
                 (EpAnn
                  (Anchor
                   { <input>:3:5 }
                   (UnchangedAnchor))
                  (AnnSig
                   (AddEpAnn AnnDcolon (EpaSpan { <input>:3:7-8 }))
                   [])
                  (EpaComments
                   [(L
                     (Anchor
                      { <input>:4:7-20 }
                      (UnchangedAnchor))
                     (EpaComment
                      (EpaDocCommentNext
                       " misplaced")
                      { <input>:3:7-8 }))]))
                 [(L
                   (SrcSpanAnn (EpAnnNotUsed) { <input>:3:5 })
                   (Unqual
                    {OccName: a}))]
                 (HsWC
                  (NoExtField)
                  (L
                   (SrcSpanAnn (EpAnnNotUsed) { <input>:5:7-9 })
                   (HsSig
                    (NoExtField)
                    (HsOuterImplicit
                     (NoExtField))
                    (L
                     (SrcSpanAnn (EpAnnNotUsed) { <input>:5:7-9 })
                     (HsTyVar
                      (EpAnn
                       (Anchor
                        { <input>:5:7-9 }
                        (UnchangedAnchor))
                       []
                       (EpaComments
                        []))
                      (NotPromoted)
                      (L
                       (SrcSpanAnn (EpAnnNotUsed) { <input>:5:7-9 })
                       (Unqual
                        {OccName: Int})))))))))])))))])
       (FromSource))
      [])))]
  (Nothing)
  (Nothing)))
```

</details>
